keps/sig-network/1860-kube-proxy-IP-node-binding/README.md

```diff
@@ -0,0 +1,113 @@
+
+## Motivation
+
+There are numerous problems with the current behaviour:
+
+1. Some cloud providers (Scaleway, Tencent Cloud, ...) are using the LB's external IP (or a private IP) as source IP when sending packets to the cluster. This is a problem in the ipvs mode of kube-proxy since the IP is bounded to an interface and healthchecks from the LB are never coming back.
+2. Some cloud providers (DigitalOcean, Scaleway, ...) have features at the LB level (TLS termination, PROXY protocol, ...). Bypassing the LB means missing these features when the packet arrives at the service (leading to protocols errors).
+
+So, giving options to these cloud providers to disable the actual behaviour would be very valuable.
+
+Currently there is some hacky workaround that set the `Hostname` on the `Service` in order to bypass kube-proxy binding (AWS and DigitalOcean for instance), but as stated, it's a bit hacky. This KEP should also give a solution to this hack.
+
+### Goals
+
+* Expose an option for the different controllers managing Load Balancer (e.g. the Cloud Controller Manager), to choose the actual behaviour of kube-proxy for all services.
+
+### Non-Goals
+
+* Deprecate the default behaviour of kube-proxy
+* Change the default behaviour of kube-proxy
+
+## Proposal
+
+The solution would be to add a new field in the `loadBalancer` field of a `Service`'s `status`, like `ingressMode`. This new field will be used by kube-proxy in order to not bind the Load Balancer's External IP to the node (in both IPVS and iptables mode). The value `VIP` would be the default one (if not set for instance), keeping the current behaviour. The value `Proxy` would be used in order to disable the the shortcut.
+
+Then, the Cloud Controller Manager interface would need a new method, that returns the value of this `ingressMode`, for a given cloud provider and a given service.
+
+i.e:
+
+```go
+GetLoadBalancerIngressMode(ctx context.Context, clusterName string, service *v1.Service) LoadBalancerIngressMode
+```
+
+Finally the cloud controller will call this method when the Service becomes a LoadBalancer service and set the new `ingressMode` field accordingly.
+
+### User Stories
+
+#### Story 1
+
+User 1 is a Managed Kubernetes user. There cluster is running on a cloud provider where the LB's behaviour matches the `VIP` `ingressMode`.
+Nothing changes for them since the cloud provider manages the Cloud Controller Manager.
+
+#### Story 2
+
+User 2 is a Managed Kubernetes user. There cluster is running on a cloud provider where the LB's behaviour matches the `Proxy` `ingressMode`.
+Almost nothing changes for them since the cloud provider manages the Cloud Controller Manager.
+The only difference is that pods using the load balancer IP may observe a higher response time since the datapath will go through the load balancer. 
+
+#### Story 3
+
+User 3 is a developer working on a cloud provider Kubernetes offering. 
```

Comment:
This needs to be something like:
```
On the next verison of `k8s.io/cloud-provider`, the cloud provider can optionally set the `ingressMode` field as part of LoadBalancerStatus.
```

Reply:
We need to find a good name.

ingressMode
deliveryMode
trafficPattern
just "mode" (sine it is under "ingress" already) ?

Reply:
Some names I've been steeping on:
* passthrough: true|false
* clusterPassthrough: true|false 
* inCluster: true|false
* routeLocal: true|false

I think "modes" in this case is confusing because we really only care about whether local routes are created in-cluster.

Reply:
Also, should the "mode" be per ingress or on the entire status?

Reply:
I am thinking of this as "metadata about the specific ingress IP".  So, it's possible to have multiple ingressen for a single service, and for them to be different.  I was thinking of it as "the mode of the IP" not "the mode of ingress".  But "mode" is an awful word that means nothing.  "type" would be just as vague.  "archetype"  or  "model"  or ???

Reply:
Okay so the new field per ingress. Should we have a boolean? Or keep it extensible with a string? 
If it's a string, for the naming maybe `routeType`?

Reply:
I think @thockin wants another iteration of this KEP where we also flip the status from `provisional` -> `implementable`. I don't think we need to agree on the name now but ideally we agree on the name when we set the KEP status to `implementable`, worse case we can agree on naming in the implementation PR.

I think we're good to merge the KEP as provisional (once the other comments are addressed) since we all agree this is an issue that needs fixing.

Reply:
:+1:

Reply:
I'm very open to names - it's well established that I suck at naming.